### ISSUE TITLE
Event/Webhook on User IDP link creation

### ISSUE THREAD

**Issue by matthewhartstonge** [FusionAuth/fusionauth-issues#1525]
### Problem
When a user has signed in to our application using a social sign in, we want to be able to display what connections have been made.

At the moment, in order to do this programmatically, we have to list all the IDPs, work out which one is which type, then query for the users links, then do some UUID interpolation to work out if they've connected to a given social network when pulling a user record. If we know the lsit, we can display the connected IDPs and also provide an escape hatch for those who want to 'disconnect' their bound social accounts from within our app.

It would be great if an event/webhook was possible on link creation, so that we only have to update our user objects _once_ when the binding occurs rather than having to query every time for bindings, or listen to every successful sign in.

### Solution

- Add an event type that fires on External IDP linking, for example, `user.externalidp.link.complete`
- Add the IdentityProvider type to the IDP "Link" model (single API query, instead of mapping between two API calls).
```json
{
  "identityProviderLinks" : [
    {
      "displayName" : "erlich@piedpiper.com",
      "identityProviderId" : "f50dbb83-4cc2-4e5e-aece-9efe068bddd9",
      "identityProviderType" : "Facebook"
      "identityProviderUserId" : "42",
      "insertInstant" :1623183147998,
      "lastLoginInstant" : 1623183152224,
      "tenantId" : "30663132-6464-6665-3032-326466613934",
      "token" : "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c",
      "userId" : "fbf0e652-b2b5-45eb-b9c9-d4640889740b"
    }
  ]
}
```

### Alternatives/workarounds

- On listing users, perform the multi-API query dance as above and a API link request per user, merging in at run time...
- listen to `user.login.success` and update local records based on every sign in users perform.

### Additional context
N/A

### Related
- https://github.com/FusionAuth/fusionauth-issues/issues/1589

### Community guidelines
All issues filed in this repository must abide by the [FusionAuth community guidelines](https://fusionauth.io/community/forum/topic/1000/code-of-conduct).

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by mooreds** [comment]
@matthewhartstonge FYI, the ids of identity providers for all social logins are fixed as documented in the various pages. 

> The Id of the Apple identity provider, which will always be 13d2a5db-7ef9-4d62-b909-0df58612e775.

https://fusionauth.io/docs/v1/tech/apis/identity-providers/apple/

So if you are only worried about social providers, you could embed a map of the id -> type in your code rather than making a second API call.

That's secondary to your request, but just wanted to note that.

**Comment by TroubleShutter** [comment]
Maybe you shouold use reconcile lambda to put IDP Link to user.data object and then use it in webhooks?
https://fusionauth.io/docs/v1/tech/lambdas/openid-connect-response-reconcile/

**Comment by glen-84** [comment]
@mooreds Should this be closed now that `user.identity-provider.link` is available?

**Comment by mooreds** [comment]
@glen-84 thanks for checking! We're still working through getting all the docs out for the 1.36 release, so we'll close this out when that is done.

**Comment by robotdan** [comment]
If this is a dup of https://github.com/FusionAuth/fusionauth-issues/issues/1589 - that one was delivered, so can we close this one?

**Comment by mooreds** [comment]
@matthewhartstonge does #1589 deliver what you need? From my reading it is a dup, but you filed the issue.

**Comment by matthewhartstonge** [comment]
Yeah, definitely looks like it. This way we can hook into the event to propagate what IDPs a user is connected to to support unlinking/connection display to our users easily.

Therefore, will be fixed via #1589.
